### ISSUE TITLE
Something error with Compiler Test Case

### ISSUE THREAD

**Issue by ooggss** [rust-lang/rust#134515]
I download the whole project(with commit `706141b8d9090228343340378b1d4a2b095fa1fb`) and try to reproduce the compile error with compiler test cases in the project. But I find that rust compiler(both `1.77.1-x86_64-unknown-linux-gnu` and `rustc 1.82.0 (f6e511eec 2024-10-15)` version) doesn't perform as expected on many of the compiler test cases.

For example, [this file](https://github.com/rust-lang/rust/blob/master/src/tools/clippy/tests/ui/semicolon_inside_block.rs) is supposed to be fail to compile and get such [error message](https://github.com/rust-lang/rust/blob/master/src/tools/clippy/tests/ui/semicolon_inside_block.stderr). But Actually I compile it successfully.

And this is not a rare case, I tried to reproduce all test cases with .fixed version(1,062 in all) and 418 of them are not as expected, i.e. they can be compiled successfully.

Is this because the data has not been updated, or is there something wrong with my operation?

**Comment by bjorn3** [comment]
How did you try to test?

> Is this because the data has not been updated

CI should fail if they are not updated.

**Comment by jieyouxu** [comment]
The test you linked is a *clippy* ui test, not a *rustc* ui test. How are you running the clippy tests?

**Comment by ooggss** [comment]
> The test you linked is a _clippy_ ui test (`src/tools/clippy/tests/ui/semicolon_inside_block.rs`), not a _rustc_ ui test. How are you running the clippy tests?

Could you explain the difference between clippy tests with rustc tests please? Does it mean that these are not compile error?

**Comment by jieyouxu** [comment]
For tests in *this* repo, they'll have to go through the build system bootstrap. See <https://rustc-dev-guide.rust-lang.org/tests/intro.html>. I don't remember if clippy tests are supported in *this* repo, because clippy is developed separately in https://github.com/rust-lang/rust-clippy, and the clippy submodule (or was it subtree) in this repo is mostly snapshots of that. If you wanted to contribute to clippy, please refer to https://github.com/rust-lang/rust-clippy.

**Comment by jieyouxu** [comment]
> Could you explain the difference between clippy tests with rustc tests please? Does it mean that these are not compile error?

Those are *clippy* ui tests, which exercise the outputs of *clippy* lints and such. *rustc* ui tests live under `tests/ui`.

**Comment by ooggss** [comment]
> > Could you explain the difference between clippy tests with rustc tests please? Does it mean that these are not compile error?
> 
> The test you linked is a _clippy_ ui test, which exercise the outputs of _clippy_ lints and such. _rustc_ ui tests live under `tests/ui`.

I got it. Thanks for your reply. I’ll check if there is the same situation under `tests/ui`

**Comment by jieyouxu** [comment]
Note that you can't just `cargo build` the `tests/ui` either, you *have* to go through the bootstrap test harness, see the link to rustc-dev-guide I sent above.

**Comment by bjorn3** [comment]
You should not ever use plain cargo commands inside the rust repo. There are a bunch of things that we need to do (like bootstrapping the compiler, building the standard library, and passing special flags in certain cases) that necessitate us having a separate build system for rustc which can invoke using `x.py`. (Internally it uses cargo for building individual components of the compiler, but passes a lot of flags and env vars that you should not manually try to reproduce.)

**Comment by ooggss** [comment]
> You should not ever use plain cargo commands inside the rust repo. There are a bunch of things that we need to do (like bootstrapping the compiler, building the standard library, and passing special flags in certain cases) that necessitate us having a separate build system for rustc which can invoke using `x.py`. (Internally it uses cargo for building individual components of the compiler, but passes a lot of flags and env vars that you should not manually try to reproduce.)

Does it mean that there is are actually some compile errors in these .rs file but they just cannot be triggered by using plain `cargo build` command?

**Comment by bjorn3** [comment]
Indeed. These errors are expected to happen when clippy built from `src/tools/clippy` gets run on the .rs file. And to build this version of clippy you need to build the rustc version in compiler/, which in turn requires building the standard library version in library/. `./x.py test src/tools/clippy` handles all building necessary to run clippy's test suite and then actually runs it.

**Comment by ooggss** [comment]
> And to build this version of clippy you need to build the rustc version in compiler/,

Thanks for your reply! I want to ask another question: Does you mean that the rustic version in compiler/ is different from the version I installed from command `rustup`? If so, which version I can install from command `rustup` is the same as the version in compiler/?

**Comment by bjorn3** [comment]
> Does you mean that the rustic version in compiler/ is different from the version I installed from command rustup?

Generally yes

> If so, which version I can install from command rustup is the same as the version in compiler/?

If you cloned the git repo for rustc, you likely got a commit which isn't available on rustup. Nightlies are only released once a day and we merge changes to the rustc repo many times a day. If you are working on rustc, you should forget about rustup entirely. `./x.py` handles getting the right toolchain for you. If you want to work on clippy however, you will want to clone https://github.com/rust-lang/rust-clippy/ instead and follow the contributing instructions for that repo.

**Comment by ooggss** [comment]
> If you are working on rustc, you should forget about rustup entirely. `./x.py` handles getting the right toolchain for you. If you want to work on clippy however, you will want to clone https://github.com/rust-lang/rust-clippy/ instead and follow the contributing instructions for that repo.

Actually I want to collect some compile error examples that developer will encounter during project development as well as the fixed example to show how to fix these syntax errors for my work.

**Comment by ooggss** [comment]
> In that case you should ignore the test suites of rustc and clippy entirely. They contain a lot of weird things that no sane person would write, yet should still be handled gracefully by the compiler.

😭that’s a bad news. Could you please recommend some useful data source like Linux kernel CVEs

**Comment by bjorn3** [comment]
Are you looking for frequently occuring mistakes people make that the compiler catches? Or ones that the compiler doesn't catch? If the latter, all mistakes or only those that could be considered security issues?

**Comment by ooggss** [comment]
> Edit: I think you are looking for the former based on your earlier comment. 

Yes it is the former one😭

Anyway, thank you very much!!

**Comment by Noratrieb** [comment]
the question has been answered, closing
